docs/proposals/2170-kubeflow-training-v2/README.md

```diff
@@ -0,0 +1,1362 @@
+  training job without understanding complex Kubernetes APIs.
+- Update Kubeflow Training SDK to allow data scientists quickly create and monitor `TrainJobs`.
+- Create community-supported `ClusterTrainingRuntime` for distributed training with PyTorch and MPI.
+- Create community-supported `ClusterTrainingRuntime` for LLM fine-tuning for various foundational
+  models (e.g. Mistral, LLama-70b, Gemma-7b).
+- Work on the following JobSet improvements: https://github.com/kubernetes-sigs/jobset/issues/463 and https://github.com/kubernetes-sigs/jobset/issues/572
+
+### Non-Goals
+
+- Support MPI V1 implementation.
+- Distributed training for TensorFlow, XGboost, JAX, and PaddlePaddle will be added after initial
+  implementation.
+- Migrate Kubeflow V1 controller to use `JobSet`.
+
+## Design Details
+
+We propose these APIs:
+
+- **`TrainJob`**: A single API which allows data scientists to initiate a training and fine-tuning
+  job from the pre-deployed training runtime. It allows users to tweak configurations for their
+  training jobs such as model parameters, dataset parameters, or trainer configuration.
+  The main goal is to hide unnecessary Kubernetes complexity for data scientists.
+
+- **`TrainingRuntime`** and **`ClusterTrainingRuntime`**: Set of blueprints for how to start various
+  types of training or fine-tuning jobs. Runtimes are managed by Platform Engineers and allow them
+  to configure infrastructure parameters that are required for the **TrainJob**.
+  For example, failure policy or gang-scheduling.
+
+The below diagram shows which resources will be created for LLM fine-tuning with PyTorch.
+
+![trainjob-diagram](./trainjob-diagram.jpg)
+
+### Worker and Node Definition
+
+To better understand what does Nodes and Worker mean in the diagram above,
+the following table explains naming that each framework or technology uses:
+
+<table>
+  <tr>
+   <td><strong>ML Framework or Technology</strong>
+   </td>
+   <td><strong>Definition of a Single Device (GPU)</strong>
+   </td>
+   <td><strong>Definition of a Single VM</strong>
+   </td>
+   <td><strong>Start Command</strong>
+   </td>
+   <td><strong>Reference Docs</strong>
+   </td>
+  </tr>
+  <tr>
+   <td>Kubernetes
+   </td>
+   <td>Container Resource Unit
+   </td>
+   <td>Pod’s Container
+   </td>
+   <td>Any
+   </td>
+   <td><a href="https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#resource-units-in-kubernetes">Resource units</a> in K8s
+   </td>
+  </tr>
+  <tr>
+   <td>PyTorch
+   </td>
+   <td>Worker 
+<p>
+<code>(--nproc-per-node)</code>
+   </td>
+   <td>Node
+<p>
+<code>(--nnodes)</code>
+   </td>
+   <td><code>torchrun</code>
+   </td>
+   <td><a href="https://pytorch.org/docs/stable/elastic/run.html">PyTorch Elastic</a>
+   </td>
+  </tr>
+  <tr>
+   <td>MPI
+   </td>
+   <td>Slot
+<p>
+<code>(--np)</code>
+   </td>
+   <td>Node 
+<p>
+<code>(--host)</code>
+   </td>
+   <td><code>mpirun</code>
```

Comment:
Thanks for sharing. Do you have any useful links that we can add here for reference to explain how different MPI implementations map workers and nodes to physical resources ?

Reply:
What I have quickly handy is a guide for resource managers, which more efficiently bootstrap MPI: https://hpc.llnl.gov/banks-jobs/running-jobs/batch-system-cross-reference-guides. I couldn't find a bare bones MPI launcher Rosetta Stone, and I started putting one together tonight for OpenMPI, MPICH, Intel MPI, and spectrum MPI. I want to sanity check a few things but will post the link in the next few days!

Reply:
Alright - here is a quick rundown, for the top set that I see most frequently. https://gist.github.com/vsoch/9ac7c4448dffe656d946edceaa58bd9e

I think to generally say `-n or -np` is the flag for processes is correct, and there is a bit more variation in the host definition (e.g., `-host` vs `-f` vs `-hostfile` for comma separated command line vs file) but for this table it would work to provide an example from one implementation (OpenMPI) as you have done, but then note that the actual interface can vary between implementations, but also within an implementation depending on your needs for topology / how you ask for resources. For example, asking for a total number of nodes (pods) with `-N` and then specifying processes per node, vs providing a hostlist and a _total_ number of processes (and letting MPI figure out the distribution) is different. Asking for a specific topology (which will give you the correct performance) is different. In HPC we typically don't interact with MPI directly, but use a workload launchers to more efficiently bootstrap it. As long as the message isn't strongly "there is one MPI, it's mpirun, and this is what it looks like" I think it's OK.

Reply:
Thanks for summarizing this! Let me add this as a reference to better understand the signature between different MPI implementations.